fix: prevent duplicate ID generation in fast multi-block renders

binary8421 opened the pull request:
## :bookmark_tabs: Summary

Fix duplicate ID generation when rendering multiple Mermaid blocks quickly.

## :straight_ruler: Design Decisions

This fix changes the ID generation to use microsecond-level timestamps instead of milliseconds, greatly reducing the chance of duplicate IDs.  
Additionally, if a duplicate ID is detected, an incremental counter is added to ensure uniqueness.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [ ] :computer: have added necessary unit/e2e tests.
- [ ] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [ ] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

binary8421 committed:
fix: prevent duplicate ID generation in fast multi-block renders

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

sidharthv96 commented on packages/mermaid/src/utils.ts:
```diff
@@ -759,7 +805,7 @@ export class InitIDGenerator {
     // v11: Use the actual value of seed string to generate an initial value for count.
     this.count = seed ? seed.length : 0;
-    this.next = deterministic ? () => this.count++ : () => Date.now();
+    this.next = deterministic ? () => this.count++ : () => TimeProvider.getTimestamp();
```
Doing this should solve the problem of duplicate ID, and we won't have to add the complexity of `TimeProvider`. 
We don't have a requirement that the ID should be the timestamp. 

```suggestion
    this.next = deterministic
      ? () => this.count++
      : () => Number.parseInt(`${Date.now()}${this.count++}`);
```

---

Or even simpler, we could concatenate it and return the string, we'd have to fix the tests to match the functionality though.

sidharthv96 replied:
Are you using AI for this? 🤔

binary8421 committed:
Update packages/mermaid/src/utils.ts

By the way, since this class is created every time, the "+1" operation has no effect.

Co-authored-by: Sidharth Vinod <github@sidharth.dev>

binary8421 commented on packages/mermaid/src/utils.ts:
```diff
@@ -805,7 +805,9 @@ export class InitIDGenerator {
-    this.next = deterministic ? () => this.count++ : () => TimeProvider.getTimestamp();
+    this.next = deterministic
+      ? () => this.count++
+      : () => Number.parseInt(`${Date.now()}${this.count++}`);
```
By the way, since this class is created every time, the "+1" operation has no effect.

sidharthv96 replied:
I think the implementation has diverged somehow. My expectation was the IdGenerator was for individual nodes of the diagram, as the doc suggests. But currently it's only used for the ID of the diagram (which makes the deterministic flag pointless, as you said).

https://github.com/mermaid-js/mermaid/blob/797ba43d6e63af44538390b93b3674302c5ed545/packages/mermaid/src/config.type.ts#L164-L174

sidharthv96 replied:
In this case, we'd be better off just using a `cuid` for diagram id. Which has timestamp and a random component.

binary8421 commented on packages/mermaid/src/utils.ts:
```diff
@@ -805,7 +805,9 @@ export class InitIDGenerator {
     // TODO: Seed is only used for length?
     // v11: Use the actual value of seed string to generate an initial value for count.
     this.count = seed ? seed.length : 0;
-    this.next = deterministic ? () => this.count++ : () => TimeProvider.getTimestamp();
```
Using a constant object to handle conflicts is a good approach.

binary8421 reviewed (commented):
I've tested your code, and ID conflicts still occur. Detailed comments have been added.

sidharthv96 commented:
@binary8421 the IDGenerator is created once per invocation of `run` function, not for every diagram. 
Are you calling `run` in parallel, what's the use case for that? `run` is expected to be used once during page load, to render all mermaid diagrams with the given selector. 

@EnochGao This PR only handles IDs for diagrams, not nodes. Can you clarify your use case as well?
